atlassian/github-for-jira
# Arc-1455 server url to app creation

**Rationale:**
Continuing task of wiring up the UI screens. This PR focuses on navigating from the GHE server URL screen to the app creation screen.

**Changes:**

- Added 2 new routes: `JiraVerifyServerUrlRouter` to handle routing a post request (`JiraVerifyServerUrlPost`) to validate the passed url is a valid url, check the db for existing apps and verify we can hit the provided url. `JiraAppCreationRouter` handles the get request after the validation passes.
- New `JiraVerifyServerUrlPost` middleware which takes in the provided url and makes a simple get request to check that we ping the provided url. If customers have opened a hole in their firewall, we should get a successful response back, otherwise, it will fail.
- From here, we either navigate to the app creation page or render the list of GH apps (this part of the flow is yet to be implemented). https://github.com/atlassian/github-for-jira/pull/1312/files#r909455400
- New module in connect and updated snapshots of the connect object.
- Modifications to static/js/jira-server-url.js to handle making requests to our new route. 
- What's not included: styling clean-up. There's a noticeable size difference in the containers between these 2 views but we can clean this up later on.

**Demo:**
Like the last 2 PRs, the new path is hidden behind a FF.

With FF set to false we still navigate from the config screen to the connect an org screen:

https://user-images.githubusercontent.com/37155488/176413772-25add353-3d21-4717-bd71-3408dc6b4368.mov

With the FF set to true (done manually on my local for demo purposes only) we can navigate from the config page to the server url screen. In the scenario where I have 'opened a hole in the firewall' (aka connected to the VPN) and I have no existing GH apps, I'm navigated to the auto/manual app creation page.

https://user-images.githubusercontent.com/37155488/176414370-83b638d5-9f3e-4e7f-aee3-b83d1e7a4210.mov

If I do have existing apps, instead of being navigated to /github-app-creation-page I'll be navigated to `github-list-apps-page` (the view and the module for this don't exist yet so it renders nothing and the error flag pops up but not the path)

https://user-images.githubusercontent.com/37155488/176415235-6f55c824-d179-4e65-8a53-2c314d5df7b6.mov

If we can't verify the url, we'll render an error ([this component hasn't been created yet](https://github.com/atlassian/github-for-jira/pull/1312/files#r909332712) but note the response in my terminal ). For demo purposes, I'm disconnected from my VPN for this one.

https://user-images.githubusercontent.com/37155488/176415869-b02ff8a6-3440-41f2-875c-5203db761ba3.mov

To test **locally:**
In src/routes/jira/configuration/jira-configuration-get.ts change the value from false to true for the FF check for handleNavigationClassName. You will be able to test these changes again once this has been deployed to prod as the FF is on for our segment 🎉 

**What's up next:**
In [ARC 1456](https://softwareteams.atlassian.net/jira/software/projects/ARC/boards/385?assignee=604894ffe19f910069497c6a&selectedIssue=ARC-1456) I'll be implementing the auto app creation flow, saving the data from the response to the db, and wrapping out the first-time install flow. I'm going to tackle a couple of clean-up items and some documentation so I can give this bad boy a chance to get approved and merged so I stop branching off my previous branch and blowing out my commit history 🙃 

Also [ARC 1473](https://softwareteams.atlassian.net/jira/software/projects/ARC/boards/385?assignee=604894ffe19f910069497c6a&selectedIssue=ARC-1473) to map errors when the verify server post fails and rendering the error in the UI.
